### PR TITLE
request improvements

### LINKED ISSUE #33: Suggestions for HttpRequest class
After merging the `HttpRequest` PR and using it in other PRs I have a couple of suggestions:

- [x] We need to make `HttpRequestImpl` independent of nodejs. Users need to create requests when using `HttpClient`. I think it is easy to move the current implementation outside of node folder. The only dependency is `import { URLSearchParams } from "url"` which is available through `window` as well.
- [x] The order of parameters can be improved. I think it would be better if we reorder parameters to have the most important params (`url` and `method`) first. `url` and `method` are mandatory parameters for a request but the other two are not. (https://github.com/http4ts/http4ts/pull/54)
- [x] Add helper functions to easily create a request object. For example compare the following syntaxes: (https://github.com/http4ts/http4ts/pull/54)
```ts
const req1 = new HttpRequestImpl("http://localhost/x", null, "GET");
// vs
const req2 = req({ url: "http://localhost/x", method: "GET" });
// or
const req3 = req("http://localhost/x", "GET");
```
- [x] Provide functions to replace all the fields. Since we are going to use the same class in `HttpClient`, the users should be able to manipulate all the fields immutably. For example, they should be able to create another request using an existing request but only change the `url`. So I think we need to add `setUrl`, `setMethod`, `setHeaders`, `setBody` methods as well. (https://github.com/http4ts/http4ts/pull/54)

### PR DESCRIPTION
**Pull request by alisabzevari**
To close #33

### CODE REVIEW

File: src/core/http-request/http-request-impl.ts [hunk 1]
```diff
@@ -29,7 +33,7 @@ export class HttpRequestImpl implements HttpRequest {
       ...this.headers,
       [header]: value
     };
-    return new HttpRequestImpl(this.url, this.body, this.method, newHeaders);
+    return new HttpRequestImpl(this.url, this.method, this.body, newHeaders);
```
Comment (burimshala) [inline src/core/http-request/http-request-impl.ts]: What are the criteria for ordering the parameters in this way?

Comment (alisabzevari) [reply src/core/http-request/http-request-impl.ts]: `url` and `method` are mandatory but a request without body or any header is still a valid request.

File: src/core/http-request/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { HttpMethod, HttpBody, HttpHeaders, HttpMethods } from "../http";
+import { stringBody, HttpRequestImpl } from "../../node";
+
+export type RequestParams = {
+  url: string;
+  method?: HttpMethod; // Method can also be optional in this function
+  body?: HttpBody;
```
Comment (burimshala) [inline src/core/http-request/helpers.ts]: This should be string | HttpBody

File: src/core/http-request/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  headers?: HttpHeaders;
+};
+
+export function req({
+  url,
+  method = HttpMethods.GET,
+  body = stringBody(""),
+  headers = {}
+}: RequestParams) {
+  return new HttpRequestImpl(url, method, body, headers);
```
Comment (burimshala) [inline src/core/http-request/helpers.ts]: We can create the string empty body if it is not provided: `const theBody = typeof body === "string" ? stringBody(body) : body`

File: src/core/http-request/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  method?: HttpMethod; // Method can also be optional in this function
+  body?: HttpBody | string;
+  headers?: HttpHeaders;
+};
+
+/**
+ * Instantiates an HttpResponse based on the provided data
+ * @param obj Request representative object. Only url is mandatory. The default HttpMethod is GET.
+ */
+export function req({
```
Comment (burimshala) [inline src/core/http-request/helpers.ts]: Are we going to keep this function with a single parameter as an object or several parameters?

Comment (alisabzevari) [reply src/core/http-request/helpers.ts]: I think we have decided to keep it like that.

**Pull request merged**